packages/sources/tradingeconomics-test/src/endpoint/price-ws.ts

```diff
@@ -0,0 +1,85 @@
+import { WebsocketReverseMappingTransport } from '@chainlink/external-adapter-framework/transports'
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+import { EndpointTypes } from './price-router'
+
+const logger = makeLogger('TradingEconomics WS Price')
+
+export interface Message {
+  s: string
+  i: string
+  pch: number
+  nch: number
+  bid: number
+  ask: number
+  price: number
+  dt: number
+  state: string
+  type: string
+  dhigh: number
+  dlow: number
+  o: number
+  prev: number
+  topic: string
+}
+
+type WSEndpointTypes = EndpointTypes & {
+  Provider: {
+    WsMessage: Message
+  }
+}
+
+export const withApiKey = (url: string, key: string, secret: string) =>
+  `${url}?client=${key}:${secret}`
+
+export const wsTransport: WebsocketReverseMappingTransport<WSEndpointTypes, string> =
+  new WebsocketReverseMappingTransport<WSEndpointTypes, string>({
+    url: (context) => {
+      const { API_CLIENT_KEY, API_CLIENT_SECRET, WS_API_ENDPOINT } = context.adapterSettings
+      return withApiKey(WS_API_ENDPOINT, API_CLIENT_KEY, API_CLIENT_SECRET)
+    },
+    handlers: {
+      message: (message) => {
+        if (!message.topic || message.topic === 'keepalive') {
+          return []
+        }
+        const pair = wsTransport.getReverseMapping(message.s)
+        if (!pair) {
+          logger.error(`Pair not found in websocket reverse map for message symbol - ${message.s}`)
+          return []
+        }
+        if (message.price === undefined) {
+          return [
+            {
+              params: pair,
+              response: {
+                errorMessage: `Tradingeconomics provided no data for ${JSON.stringify(pair)}`,
```

Comment:
It may be helpful also to log this message. Since it's a provider issue, we can just mark it as `info`.